builtin/providers/aws/resource_aws_codebuild_project.go

```diff
@@ -0,0 +1,722 @@
+			},
+			"service_role": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"source": &schema.Schema{
+				Type: schema.TypeSet,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"auth": &schema.Schema{
+							Type: schema.TypeSet,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"resource": {
+										Type:     schema.TypeString,
+										Optional: true,
+									},
+									"type": {
+										Type:         schema.TypeString,
+										Required:     true,
+										ValidateFunc: validateAwsCodeBuildSourceAuthType,
+									},
+								},
+							},
+							Optional: true,
+						},
+						"buildspec": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"location": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"type": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validateAwsCodeBuildSourceType,
+						},
+					},
+				},
+				Required: true,
+				MaxItems: 1,
+			},
+			"timeout": {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validateAwsCodeBuildTimeout,
+			},
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsCodeBuildProjectCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).codebuildconn
+
+	log.Printf("[DEBUG] CodeBuild Create Project: %s", d.Id())
+
+	projectEnv := expandProjectEnvironment(d)
+	projectSource := expandProjectSource(d)
+	projectArtifacts := expandProjectArtifacts(d)
+
+	params := &codebuild.CreateProjectInput{
+		Environment: projectEnv,
+		Name:        aws.String(d.Get("name").(string)),
+		Source:      &projectSource,
+		Artifacts:   &projectArtifacts,
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		params.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("encryption_key"); ok {
+		params.EncryptionKey = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("service_role"); ok {
+		params.ServiceRole = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("timeout"); ok {
+		params.TimeoutInMinutes = aws.Int64(int64(v.(int)))
+	}
+
+	var resp *codebuild.CreateProjectOutput
+	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+		var err error
+
+		resp, err = conn.CreateProject(params)
+
+		if err != nil {
+			return resource.RetryableError(err)
+		}
+
+		return resource.NonRetryableError(err)
+	})
+
+	if err != nil {
+		return fmt.Errorf("[ERROR] Error creating CodeBuild project: %s", err)
+	}
+
+	if resp.Project.Name == nil {
+		return fmt.Errorf("[ERROR] Project name was nil")
+	}
+
+	d.SetId(*resp.Project.Name)
+
+	return resourceAwsCodeBuildProjectUpdate(d, meta)
+}
+
+func expandProjectArtifacts(d *schema.ResourceData) codebuild.ProjectArtifacts {
+	configs := d.Get("artifacts").(*schema.Set).List()
+	projectArtifacts := codebuild.ProjectArtifacts{}
+
+	for _, configRaw := range configs {
```

Comment:
IF artifacts has MaxItems of 1, I don't think we need to range here. We can use configs[0] and grab the items directly